File: bindings/java/Dashem.java
/**
 * Enterprise-Grade Em-Dash Removal Library for Java
 *
 * This class provides Java bindings to the high-performance C library
 * for removing em-dashes (U+2014) from UTF-8 encoded strings.
 */

public class Dashem {
    static {
        System.loadLibrary("dashem");
    }

    /**
     * Remove all em-dashes from the input string.
     *
     * @param input Input string
     * @return String with em-dashes removed
     * @throws UnsatisfiedLinkError if native library is not available
     */
    public static native String remove(String input);

    /**
     * Get library version.
     *
     * @return Version string
     */
    public static native String version();

    /**
     * Get implementation name.
     *
     * @return Implementation name (e.g., "AVX2", "SSE4.2")
     */
    public static native String implementationName();

    /**
     * Detect available CPU features.
     *
     * @return Bitmask of CPU features
     */
    public static native int detectCPUFeatures();

    public static void main(String[] args) {
        System.out.println("dash-em " + version());
        System.out.println("Implementation: " + implementationName());

        if (args.length > 0) {
            String result = remove(args[0]);
            System.out.println("Result: " + result);
        }
    }
}

File: bindings/swift/Dashem.swift
import Foundation

/**
 * Enterprise-Grade Em-Dash Removal Library for Swift
 *
 * High-performance, SIMD-accelerated string processing for removing
 * em-dashes (U+2014) from UTF-8 encoded text.
 */

@_silgen_name("dashem_remove")
private func dashem_remove(
    _ input: UnsafePointer<UInt8>,
    _ inputLen: Int,
    _ output: UnsafeMutablePointer<UInt8>,
    _ outputCapacity: Int,
    _ outputLen: UnsafeMutablePointer<Int>
) -> Int32

@_silgen_name("dashem_version")
private func dashem_version() -> UnsafePointer<CChar>

@_silgen_name("dashem_implementation_name")
private func dashem_implementation_name() -> UnsafePointer<CChar>

/// Remove em-dashes from a string
public func removeEmDashes(_ input: String) -> String {
    let inputData = input.utf8
    var outputBuffer = [UInt8](repeating: 0, count: inputData.count)
    var outputLen: Int = 0

    let result = inputData.withContiguousStorageIfAvailable { inputBuffer in
        return dashem_remove(
            inputBuffer.baseAddress!,
            inputBuffer.count,
            &outputBuffer,
            outputBuffer.count,
            &outputLen
        )
    }

    guard result == 0 else {
        fatalError("dashem_remove failed")
    }

    return String(bytes: outputBuffer[..<outputLen], encoding: .utf8) ?? ""
}

/// Get library version
public func version() -> String {
    return String(cString: dashem_version())
}

/// Get implementation name
public func implementationName() -> String {
    return String(cString: dashem_implementation_name())
}
